src/client/dfuse/ops/readdir.c

```diff
@@ -24,240 +24,156 @@
 #include "dfuse_common.h"
 #include "dfuse.h"
 
-/* TODO: This implementation is not complete, in particular it does not
- * correctly handle calls where offset != 0, which potentially generates
- * incorrect results if the number of files is greater than can be observed in
- * one call (approx 25).
- *
- * If the filesystem is only modified from one client then the results should
- * however be correct.
- */
+#define LOOP_COUNT 20
+
+struct iterate_data {
+	fuse_req_t			req;
+	struct dfuse_inode_entry	*inode;
+	struct dfuse_obj_hdl		*oh;
+	void				*buf;
+	size_t				size;
+	size_t				b_offset;
+};
+
+int
+filler_cb(dfs_t *dfs, dfs_obj_t *dir, const char name[], void *_udata)
+{
+	struct iterate_data	*udata = (struct iterate_data *)_udata;
+	struct dfuse_projection_info *fs_handle = fuse_req_userdata(udata->req);
+	struct dfuse_obj_hdl	*oh = udata->oh;
+	dfs_obj_t		*obj;
+	daos_obj_id_t		oid;
+	struct stat		stbuf = {0};
+	int			ns = 0;
+	int			rc;
+
+	/*
+	 * MSC - from fuse fuse_add_direntry: "From the 'stbuf' argument the
+	 * st_ino field and bits 12-15 of the st_mode field are used. The other
+	 * fields are ignored." So we only need to lookup the entry for the
+	 * mode.
+	 */
+	DFUSE_TRA_DEBUG(udata->inode, "Adding entry name '%s'", name);
+
+	rc = dfs_lookup_rel(dfs, dir, name, O_RDONLY, &obj, &stbuf.st_mode);
+	if (rc)
+		return rc;
+
+	rc = dfs_obj2id(obj, &oid);
+	if (rc)
+		D_GOTO(out, rc);
+
+	rc = dfuse_lookup_inode(fs_handle, udata->inode->ie_dfs, &oid,
+				&stbuf.st_ino);
+	if (rc)
+		D_GOTO(out, rc);
+
+	/* if the buffer on the OH has not been used yet */
+	if (oh->doh_buf == NULL) {
+		/** add the entry to the buffer to return to fuse first */
+		ns = fuse_add_direntry(udata->req, udata->buf + udata->b_offset,
+				       udata->size - udata->b_offset, name,
+				       &stbuf, (off_t)(&oh->doh_anchor));
+		/** If entry does not fit */
+		if (ns >= udata->size - udata->b_offset) {
+			/** alloc a buffer on oh to hold further entries */
+			D_ALLOC(oh->doh_buf, udata->size);
+			if (!oh->doh_buf)
+				return -ENOMEM;
```

Comment:
should probably call dfs_release()?
Actually, the object could probably be released just after dfs_obj2id().

Reply:
ah right.. since you merged already ill fix in next patch